programs/open_creator_protocol/src/instructions/nft_proxy/migrate_to_mpl.rs

```diff
@@ -0,0 +1,157 @@
+        seeds::program = anchor_spl::metadata::Metadata::id(),
+        bump,
+    )]
+    metadata: Box<Account<'info, MetadataAccount>>,
+    #[account(
+        mut,
+        close = from,
+        seeds = [MintState::SEED.as_bytes(), mint.key().as_ref()],
+        bump,
+    )]
+    mint_state: Box<Account<'info, MintState>>,
+    #[account(mut)]
+    from: Signer<'info>, // this is the update_authority of the metadata account
+    #[account(
+        mut,
+        constraint = from_account.mint == mint.key() @ OCPErrorCode::InvalidTokenAccount,
+        constraint = from_account.amount == 1 @ OCPErrorCode::InvalidTokenAccount,
+        // for from_account.owner, we don't need to check the owner of the token account because
+        // this migration is triggered by the update_authority (i.e. the "from" account)
+    )]
+    from_account: Box<Account<'info, TokenAccount>>,
+    /// CHECK: going to create this account in cpi
+    #[account(mut)]
+    edition: UncheckedAccount<'info>,
+    /// CHECK: going to create this account in cpi
+    token_program: Program<'info, Token>,
+    system_program: Program<'info, System>,
+    /// CHECK: checked in cpi
+    #[account(address = community_managed_token::id())]
+    cmt_program: UncheckedAccount<'info>,
+    /// CHECK: This is not dangerous because the ID is checked with instructions sysvar
+    #[account(address = sysvar::instructions::id())]
+    instructions: UncheckedAccount<'info>,
+    /// CHECK: This is not dangerous because the ID is checked with mpl_token_metadata::id()
+    #[account(address = anchor_spl::metadata::Metadata::id())]
+    metadata_program: UncheckedAccount<'info>,
+}
+
+impl From<&mut MigrateToMplCtx<'_>> for ActionCtx {
+    fn from(ctx: &mut MigrateToMplCtx) -> Self {
+        let mut action_ctx = ActionCtx {
+            action: "migrate_to_mpl".to_string(),
+            program_ids: vec![],
+            last_memo_data: None,
+            last_memo_signer: None,
+            payer: None,
+            from: Some(ctx.from.key().to_string()),
+            to: None,
+            mint: ctx.mint.key().to_string(),
+            metadata: Some(ctx.metadata.clone().into()),
+            mint_account: None,
+            mint_state: ctx.mint_state.clone().into_inner().into(),
+        };
+        action_ctx
+            .parse_instructions(&ctx.instructions)
+            .expect("failed to parse sysvar instructions");
+        action_ctx
+    }
+}
+
+pub fn handler<'info>(ctx: Context<'_, '_, '_, 'info, MigrateToMplCtx<'info>>) -> Result<()> {
+    let action_ctx: ActionCtx = ctx.accounts.into();
+    ctx.accounts.policy.matches(&action_ctx)?;
+
+    // We need to verify that the metadata account's update authority is the same as the from account
+    // Also, we already verified the mint -> metadata derivation via the to_metadata_ctx function
+    if action_ctx.metadata.expect("metadata account should be available").update_authority != ctx.accounts.from.key().to_string() {
```

Comment:
could this be done as a constraint on `ctx.accounts.metadata`?

Reply:
updated to anchor constraint